feat: require signatures to prove control of `signer-key` in pox

hstove opened the pull request:
This closes #4247, which is to ensure that all Signers are using signing keys that they either control, or are controlled by the account they're delegating to.

- closes #4247 

This is done by adding a `signer-sig` param to:

- stack-stx
- delegate-stack-stx
- stack-aggregation-commit

The signature follows [SIP018](https://github.com/stacksgov/sips/blob/main/sips/sip-018/sip-018-signed-structured-data.md) for signed structured data. Following SIP018 has the benefit of built-in wallet support, so signers can use hardware wallets and apps to generate these signatures. 

Following SIP018, the structured data is `{ reward-cycle, reward-cycle }`, where `reward-cycle` is the _current_ stacking cycle where the transaction is made. The `domain` is `{ app: "pox-4-signer", version: "1.0.0", chain-id }`.

Because we'll almost certainly be building the ability for `stacks-signer` to generate these signatures (ie via CLI), I added a SIP18 util library at `/stacksib/src/lib_util/signed_structured_data`. That library has tests that validate the functionality based on the test vectors in SIP018.

I added a function `to_rsv` to `MessageSignature` to match the byte order that Clarity uses for signatures.

Note that I added the new `signer-sig` param as the second-to-last argument because of other conversations around expected argument order, but I'm not really familiar with the requirements there. I also am not sure what other touch points use these functions where the arguments need to be updated. cc @zone117x regarding param order

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -658,6 +664,28 @@
+    (
+      (msg { stacker: stacker, reward-cycle: (current-pox-reward-cycle) })
+      (msg-bytes (unwrap! (to-consensus-buff? msg) ERR_DELEGATION_INVALID_SIGNATURE)) ;;TODO
+      (msg-hash (sha256 msg-bytes))
```
We use sha512/256 for our message digests

hstove replied:
As in sha512(sha256(message))? No problem, I'll change that.

hstove replied:
I also notice that in [SIP018](https://github.com/stacksgov/sips/blob/main/sips/sip-018/sip-018-signed-structured-data.md#formal-specification) (signed structured data) only `sha256` is used, and I'd like to build this off of SIP018.

hstove replied:
The latest in this PR does follow SIP18, the v1 definitely didn't

hstove commented:
Do we need to add the signature check to `delegate-stack-increase` and `delegate-stack-extend`?

hstove commented:
After further thought, I think we should follow SIP018 (signed structured data) for this. That also unlocks the ability for signers to use a web app to request signatures from a wallet.

friedger commented:
#4247 is about delegating signer duties, not delegating stx. As far as I understand. 

Pool operators can already refuse to `delegate-stack-stx` for a pool member.

hstove commented:
@friedger @kantai is it better to describe this ticket as "delegates and delegators need to prove ownership of the signing key"? For delegates, this is proving that the signer they're delegating to has control over the signing key.

jcnelson commented:
_Everyone_ needs to prove ownership of their signing key.  This goes for solo stackers as well, via `stack-stx` and `stack-extend`.  Basically, anywhere you see a signing key as an argument, you need to also require a signature argument and a signature verification (that should be `stack-stx`, `delegate-stack-stx`, `stack-extend`, and `delegate-stack-extend`).

friedger commented:
Instead of delegate stack stx, pool operators now provide the key in stack aggregation-* function calls.

See #4274

hstove commented:
> Instead of delegate stack stx, pool operators now provide the key in stack aggregation-* function calls.
> 
> See #4274

Good to know - and I'll have to change this to build off the change from `signer-key` to `signer` too.

hstove commented:
> _Everyone_ needs to prove ownership of their signing key. This goes for solo stackers as well, via `stack-stx` and `stack-extend`. Basically, anywhere you see a signing key as an argument, you need to also require a signature argument and a signature verification (that should be `stack-stx`, `delegate-stack-stx`, `stack-extend`, and `delegate-stack-extend`).

👍 that's reflected in the current state of the PR

hstove commented:
There are a number of tests outside of the pox-4 tests related to nakamoto that are failing because of the change to `stack-stx`, I'm working on those still but the pox-4 code is ready for review.

zone117x commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -569,6 +570,7 @@
                           (start-burn-ht uint)
                           (lock-period uint)
+                          (signer-sig (buff 65))
                           (signer-key (buff 33)))
```
Thoughts on removing the `signer-key` arg? Seems redundant now that `signer-sig` is passed and the key can be derived from it? I'm guessing the burn-op versions of these functions will do that to save OP_RETURN space.

hstove replied:
Interesting thought. If it weren't for the burn op issue I'd probably push back, because it's too easy for a mis-created signature to cause an invalid signing key to be used. If the message hash differs, then the signer-key will end up as something else.

jcnelson commented on stacks-common/src/util/secp256k1.rs:
```diff
@@ -109,6 +109,11 @@ impl MessageSignature {
     }
+
+    /// Convert from VRS to RSV
+    pub fn to_rsv(&self) -> Vec<u8> {
```
Why is this conversion necessary?

hstove replied:
Because when recovering or verifying signatures in Clarity, the signature needs to be in RSV order. Whether this helper function is necessary or not is a fine question, it was just helpful to re-use

jcnelson replied:
Ugh I just noticed that `secp256k1_recover()` and `secp256k_verify()` don't behave the same as the corresponding methods in `Secp256k1PublicKey` and `MessageSignature`.  The former two methods, which Clarity uses, were added by an itinerant 3rd party contributor who couldn't be bothered to figure out how the existing code worked.

hstove replied:
Yeah, that's unfortunate because almost every `sign()` implementation returns the signature in RSV, but Clarity uses VRS, so you always have to build a helper method for it. Easy to get tripped up on.

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -672,6 +678,44 @@
+      (domain-hash (sha256 (unwrap-panic (to-consensus-buff? domain))))
+    )
+    (sha256 (concat
+      0x534950303138
```
This should be a constant somewhere

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:
```diff
@@ -1235,13 +1263,15 @@ fn pox_4_revoke_delegate_stx_events() {
         bob_principal,
         None,
         None,
+        // &signature,
```
Should this be commented?

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:
```diff
@@ -1421,33 +1633,34 @@ fn stack_stx_signer_key() {
-    .unwrap();
-    let txs = vec![make_pox_4_contract_call(
-        stacker_key,
+    // let pox_addr = make_pox_addr(
```
Please drop commented code

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:
```diff
@@ -1910,14 +2141,28 @@ fn delegate_stack_increase() {
         pox_addr.clone(),
         block_height as u128,
         lock_period,
+        signature,
```
Can you add some test coverage to this file for all four affected stacking functions to verify that they fail with the expected error code if the signature was bad?  Not just unit-testing the signature verification function; verifying end-to-end that the stacking function fails as expected.  Thanks!

hstove replied:
Added 4 new tests that cover all the stacking transactions

jcnelson commented on stackslib/src/util_lib/signed_structured_data.rs:
```diff
@@ -0,0 +1,122 @@
+use clarity::vm::Value;
```
Missing copyright comment block

jcnelson commented on stackslib/src/util_lib/signed_structured_data.rs:
```diff
@@ -0,0 +1,122 @@
+
+/// Generate a message hash for signing structured Clarity data.
+/// Reference [SIP018](https://github.com/stacksgov/sips/blob/main/sips/sip-018/sip-018-signed-structured-data.md) for more information.
+pub fn structured_data_message_hash(structured_data: Value, domain: Value) -> Sha256Sum {
```
Why does this belong in `util_lib`?  Isn't this only used in testing?  If so, then please remove this file altogether and put this into `stackslib/src/chainstate/stacks/boot/mod.rs`.

hstove replied:
At the moment it's only used in tests, but we'll add it (in a separate PR) to `stacks-signer` so we can expose a CLI function like:

```
stacks-signer create-stacking-signature --stacker S --reward-cycle R
```

I don't have a strong opinion whether it should live here or not, and I'm not the best person to know. But I think it is helpful to expose somewhere that isn't just for tests.

jcnelson replied:
Sounds like this belongs in `libsigner` then.  `libsigner` can be a dev-dependency for `stackslib`.

setzeus reviewed (changes requested):
Generally pox_4.clar updates LGTM minus a few syntax issues & one ERR constant inconsistency.

hstove commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -30,6 +30,7 @@
 (define-constant ERR_INVALID_SIGNER_KEY 32)
 (define-constant ERR_REUSED_SIGNER_KEY 33)
 (define-constant ERR_DELEGATION_ALREADY_REVOKED 34)
+(define-constant ERR_DELEGATION_INVALID_SIGNATURE (err 35))
```
ah, my habits slipped in here 😬

hstove commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -672,6 +681,44 @@
+      (msg-hash (get-signer-key-message-hash stacker))
+      (pubkey (unwrap! (secp256k1-recover? msg-hash signer-sig) ERR_DELEGATION_INVALID_SIGNATURE)) ;; TODO
+    )
+    (asserts! (is-eq pubkey signer-key) ERR_DELEGATION_INVALID_SIGNATURE)
```
Good call, I also renamed the error message to be more clear (and removed `delegation` as its inaccurate)

hstove commented:
FYI I'm going to rebase/build this off of #4269, because that changes the function signatures in pox-4. Shouldn't be too much work and then I'll move the ticket back to review.

setzeus commented:
@hstove copy, will re-review once re-based.

hstove commented:
Ready for review. This was rebased off of #4269 and now off of `next`, which changed a few things from last week's version:

- the `signer-key` (and thus `signer-sig`) args were removed from `delegate-stack-*`
- `signer-sig` was added to `stack-aggregation-commit`
- Now the structured data tuple is `{ pox-addr, reward-cycle }`. Previously it was `{ stacker, reward-cycle }`

friedger commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -725,17 +775,19 @@
 ;; Returns (err ...) on failure.
 (define-public (stack-aggregation-commit (pox-addr { version: (buff 1), hashbytes: (buff 32) })
```
This method should be removed. It was kept for backwards compatibility. Now, it is not compatible anymore, therefore, we only should use `stack-aggregation-commit-indexed`

hstove replied:
I'm curious how many apps use this function vs `-indexed`. My guess is that the easiest path forward is to remove `-indexed` but change the function's return signature here.

For example, looks like the stacks.js package uses `stack-aggregation-commit`.

friedger replied:
Stacks.js has not yet been updated to stacks 2.4 🙃

Either is fine to me. Jude might have a stronger opinion.

jcnelson replied:
It's fine to remove it for pox-4, but please add a changelog entry.

friedger commented on stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:
```diff
@@ -1805,11 +2322,20 @@ fn delegate_stack_stx_extend_signer_key() {
 
-    let next_reward_cycle = 1 + burnchain
+    let stacking_state = get_stacking_state_pox_4(&mut peer, &latest_block, &alice_principal)
+        .expect("No stacking state, stack-stx failed")
```
bob called delegate-stack-stx that failed here.

Should be changed in line 2323 as well.

hstove replied:
Can you explain more? When I run this test every contract call returns `ok`, including `delegate-stack-stx`

friedger reviewed (dismissed):
The burn chain tx for stack-stx and stack-extend need to be updated.

Otherwise, just small nits. This looks good if all tests pass.

hstove commented:
> The burn chain tx for stack-stx and stack-extend need to be updated.

We have a ticket for that at #4285, should we wait for that decision to be made before merging this? I'm not sure if we have tests for those burn ops in pox-4. I think it's already broken from just `signer-key` being added, though I might be mistaken.

zone117x commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -571,6 +576,7 @@
                           (pox-addr (tuple (version (buff 1)) (hashbytes (buff 32))))
                           (start-burn-ht uint)
                           (lock-period uint)
+                          (signer-sig (buff 65))
```
Synthetic pox events are generated using argument indexes rather than named values, so it's generally easier to add new args to the end of the list. Not a deal breaker in this PR because `signer-sig` isn't yet added to the pox events, but I'm working on adding that and our PRs will conflict due to the ordering.

zone117x replied:
Conflict PR: https://github.com/stacks-network/stacks-core/pull/4329

I can modify my PR and change the event arg indexes, no problem. Just an FYI

hstove replied:
No problem, let's merge your PR and I'll update mine.

hstove replied:
I have no real preference on arg order. Is your PR the only real place we need to worry about arg index? I'd probably just update that synthetic events code to change the signer_key index, but if there are other downstream issues then it's probably better for me to change the argument order on pox-4.

zone117x replied:
Thanks!

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -670,6 +680,44 @@
+;; Note that `reward-cycle` corresponds to the _current_ reward cycle,
+;; not the reward cycle at which the delegation will start.
+;; The public key is recovered from the signature and compared to `signer-key`.
+(define-read-only (verify-signer-key-sig (pox-addr { version: (buff 1), hashbytes: (buff 32) })
```
Similarly, can this take the reward cycle as an argument, and pass it into `get-signer-key-message-hash`?

jcnelson replied:
Also, given the way this is used in `stack-aggregation-commit-indexed`, you'd actually need a caller-supplied reward cycle, since it'll be different than the current reward cycle.

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -692,6 +741,7 @@
     ;; must be called directly by the tx-sender or by an allowed contract-caller
     (asserts! (check-caller-allowed)
               (err ERR_STACKING_PERMISSION_DENIED))
+    (try! (verify-signer-key-sig pox-addr signer-sig signer-key))
```
You can use `(current-reward-cycle)` here as the reward cycle arg.

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -692,6 +741,7 @@
     ;; must be called directly by the tx-sender or by an allowed contract-caller
     (asserts! (check-caller-allowed)
               (err ERR_STACKING_PERMISSION_DENIED))
+    (try! (verify-signer-key-sig pox-addr signer-sig signer-key))
```
You'd want to pass in the given `reward-cycle` here

hstove replied:
To maintain the behavior of "signer key signatures must use _current_ reward cycle", I've passed in `(- reward-cycle u1)` into the signature validation function

jcnelson commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -670,6 +680,44 @@
+;; The message hash follows SIP018 for signing structured data. The structured data
+;; is the tuple `{ pox-addr: { version, hashbytes }, reward-cycle }`. The domain is
+;; `{ name: "pox-4-signer", version: "1.0.0", chain-id: chain-id }`.
+(define-read-only (get-signer-key-message-hash (pox-addr { version: (buff 1), hashbytes: (buff 32) }))
```
Given how this is reached from `stack-aggregation-commit-indexed`, you'll want the reward cycle to be an argument anyway -- the reward cycle won't be equal to `current-pox-reward-cycle`.

jcnelson replied:
Actually, I think we're missing some data here.  The signature should really be over the PoX address, the first reward cycle for which the signer will sign, and the number of reward cycles for which the signer will sign.  Without that last parameter, someone could use a signer's key for 12 cycles while the signer might have only wanted to sign for 1 cycle.

* For `stack-stx`, the signature would include `lock-period`
* For `stack-extend`, the signature would include `extend-count`
* For `stack-aggregation-commit-indexed`, the lockup period would just be 1.

jcnelson reviewed (dismissed):
I think we want the signature to cover some more data than it does right now.

zone117x reviewed (approved):
I'm approving the pox4 function signature and synthetic event changes, thanks @hstove!

At least 2 others should approve the pox4 behavioral changes because I'm not in a position to validate those changes.

hstove commented:
I've updated this PR with changes since receiving Jude's feedback. Tagging @jcnelson @setzeus @MarvinJanssen @friedger for reviews.

- The signature message hash now includes two new fields
  - `topic`: one of `stack-stx`, `stack-extend`, or `agg-commit`. This prevents re-using signatures across methods
  - `period`: a `uint` reflecting `lock-period` for stacking or `extend-count` for extending. For `stack-aggregation-commit`, this value is `u1`.
- The `get-signer-key-message-hash` now takes all hash fields as args, instead of dynamically calling `get-current-reward-cycle`. This allows signers to easily call this read-only method to generate a message hash.
- In `stack-aggregation-commit`, the message hash uses the reward cycle provided as user input instead of the current reward cycle.

I've also added some code in `util_lib::signed_structured_data::pox4` to allow re-using the Rust code for generating these signatures instead of re-writing a helper function in different places.

kantai reviewed (commented):
This LGTM -- I'd just suggest using the `define_named_enum!` macro for the topics, and you should add another unit test that asserts equality between the rust message hash constructor and the contract's read-only fn.

hstove commented on stackslib/src/util_lib/signed_structured_data.rs:
```diff
@@ -0,0 +1,235 @@
+        make_structured_data_domain("pox-4-signer", "1.0.0", chain_id)
+    }
+
+    pub fn make_pox_4_signer_key_message_hash(
```
:+1: Added in https://github.com/stacks-network/stacks-core/pull/4277/commits/dbacfd06c2c19f91ed531aeac0723cdb02aec248

hstove commented on stackslib/src/chainstate/stacks/boot/pox-4.clar:
```diff
@@ -670,6 +680,57 @@
+                                               (reward-cycle uint)
+                                               (topic (string-ascii 12))
+                                               (period uint))
+  (let
```
Yep good call. I actually just stripped out the entire `let` and have one statement for both methods. The `let` made writing the code easier, but there was no variable re-use so it wasn't necessary. Plus we can save a teeny bit of cost without it!

kantai commented on stackslib/src/util_lib/signed_structured_data.rs:
```diff
@@ -0,0 +1,407 @@
+        AggregationCommit("agg-commit"),
+        StackExtend("stack-extend"),
+    });
+    pub enum Pox4SignatureTopicOld {
+        StackStx,
+        AggregationCommit,
+        StackExtend,
+    }
+
```
Can `Pox4SignatureTopicOld` be removed?

kantai reviewed (approved):
LGTM -- just see if the `...Old` enum can be removed.

hstove commented:
> LGTM -- just see if the ...Old enum can be removed.

Ah crap, thanks for catching that. CI should pass and then I think this is good to merge!

blockstack-devops commented:
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
